File: codes/import-sampling.py
#!/usr/bin/env python

"""
Importance sampling
"""

import numpy as np
import scipy.stats

def importance_sampling(func, proposal_dist, p=None, sample_size=1000, bias=False):
    """Generate samples from the proposal distribution
    
    Args:
        func (function): the integrated function
        proposal_dist (scipy distribution): the proposal distribution for IS
        p (None, optional): a pdf
        sample_size (int, optional): size of sample
        bias (bool, optional): bias or unbias
    
    Returns:
        TYPE: Description
    """

    samples = proposal_dist.rvs(size=sample_size)

    weights = 1 / proposal_dist.pdf(samples) 
    if p is not None:
        weights *= p(samples)
        
    if bias:
        weights /= np.sum(weights)
        return np.sum(weights * func(samples))
    else:
        return np.mean(weights * func(samples))


if __name__ == '__main__':

    import scipy.integrate as spi
    
    def f(x):
        return np.sin(x*10) * (0<x) * (x<np.pi)

    # Define the proposal distribution (e.g., Gaussian)
    proposal_dist = scipy.stats.norm(loc=0, scale=1)

    # Perform importance sampling
    import time
    time1 = time.perf_counter()
    estimated_integral = importance_sampling(f, proposal_dist, sample_size=2000)
    time2 = time.perf_counter()
    integral, error = spi.quad(f, 0, np.pi)
    time3 = time.perf_counter()

    print(f"Importance Sampling: {estimated_integral}, Time: {time2-time1}")
    print(f"Numerical Integration: {integral}, Time: {time3-time2}")

File: codes/import_sampling.nim
#!/usr/bin/env nim

# Importance Sampling by Nimble

import random
import sequtils

import alea

type
  Func = proc (x: float): float  # float -> float
  ProposalDist = proc (): float
  Pdf = proc (x: float): float

proc sum(arr: seq[float]): float =
  var result = 0.0
  for item in arr:
    result += item
  return result

proc mean(arr: seq[float]): float =
  result = arr.sum() / float(arr.len)


proc importance_sampling(function: Func, proposal_dist: ProposalDist, p: Pdf = nil, sample_size: int = 1000, bias: bool = false): float =
  
  #[
  Importance sampling to integrate the function `function` using `proposal_dist`.
  If `bias` is `true`, the proposal distribution density `p` is required.
  ]#

  var
    samples: seq[float] = @[]
    weights: seq[float] = @[]
    w: float

  for _ in 0..<sample_size:
    let sample = proposal_dist()
    # samples.add(sample)
    if not bias:
      assert p != nil
      w = function(sample) / p(sample)
    else:
      w = function(sample)

  weights.add(w)

  if not bias:
    return weights.mean
  else:
    return weights.sum / sample_size.float

# Example usage:

# Define the function to integrate.
proc myFunc(x: float): float =
  return x * x  # Example: x^2

# Define the proposal distribution (uniform distribution in this case).
proc proposalDist(): float =
  return rand(1.0)

# Define the probability density function of the proposal distribution.
proc proposalPdf(x: float): float =
  return 1.0  # Uniform distribution on [0, 1] has constant density 1.

# Perform importance sampling.
let result = importance_sampling(myFunc, proposalDist, proposalPdf, sample_size = 10000)
echo "Estimated integral (bias corrected): ", result

# Perform importance sampling without bias correction (using uniform distribution).
let result_bias = importance_sampling(myFunc, proposalDist, sample_size = 10000, bias = true)

echo "Estimated integral (biased): ", result_bias

